File: an2md.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import re
from lxml import etree
from markdownify import markdownify as md

# 對齊原版的英數偵測
EN_REGEX = re.compile(r'^[a-zA-Z0-9$@$!%*?&#^\-_. +]+$')

def check_en(s: str) -> bool:
    return EN_REGEX.match(s or "") is not None

def element_children(el):
    """只取元素型態的小孩（略過文字節點），維持與 libxmljs.child() 類似的語意。"""
    return [c for c in el if isinstance(c.tag, str)]

def tostring(el) -> str:
    """把元素含子樹序列化成文字（給 HTML->MD 轉換用）。"""
    if el is None:
        return ""
    return etree.tostring(el, encoding="unicode")

def an2md(xml_root: etree._Element, out_path: str) -> None:
    # namespace-agnostic 查找，對齊 //debateSection 與 //heading 的語意
    debate_matches = xml_root.xpath('//*[local-name()="debateSection"]')
    if not debate_matches:
        raise ValueError("找不到 <debateSection>")
    debate = debate_matches[0]

    heading_el_list = debate.xpath('.//*[local-name()="heading"]')
    heading_text = (heading_el_list[0].text or "").strip() if heading_el_list else ""
    colon = ":" if check_en(heading_text) else "："

    md_out = f"# {heading_text}\n\n"

    # 逐一處理 debateSection 的直屬子元素
    for child in element_children(debate):
        tag = etree.QName(child).localname
        if tag == "speech":
            # 取發言者
            speaker = (child.get("by") or "").replace("#", "")
            # 處理所有 p 元素，對齊原程式但支援多段落
            kids = element_children(child)
            if kids:
                # 如果有子元素，處理所有 p 元素
                p_elements = [k for k in kids if etree.QName(k).localname == "p"]
                if p_elements:
                    content_parts = []
                    for p_el in p_elements:
                        content_html = tostring(p_el)
                        content_md = md(content_html).strip()
                        # 移除連結前後的不必要換行
                        content_md = re.sub(r'\n+(\[.*?\]\(.*?\))', r' \1', content_md)
                        content_md = re.sub(r'(\[.*?\]\(.*?\))\n+', r'\1 ', content_md)
                        # 移除斜體文字前後的不必要換行
                        content_md = re.sub(r'\n+(\*.*?\*)', r' \1', content_md)
                        content_md = re.sub(r'(\*.*?\*)\n+', r'\1 ', content_md)
                        if content_md:
                            content_parts.append(content_md)
                    content_md = "\n\n".join(content_parts)
                else:
                    # 如果沒有 p 元素，使用原邏輯
                    content_el = kids[1] if len(kids) > 1 else kids[0]
                    content_html = tostring(content_el)
                    content_md = md(content_html).strip()
                    # 移除連結前後的不必要換行
                    content_md = re.sub(r'\n+(\[.*?\]\(.*?\))', r' \1', content_md)
                    content_md = re.sub(r'(\[.*?\]\(.*?\))\n+', r'\1 ', content_md)
                    # 移除斜體文字前後的不必要換行
                    content_md = re.sub(r'\n+(\*.*?\*)', r' \1', content_md)
                    content_md = re.sub(r'(\*.*?\*)\n+', r'\1 ', content_md)
            else:
                # 如果沒有子元素，使用文字內容
                content_md = md(child.text or "").strip()
                # 移除連結前後的不必要換行
                content_md = re.sub(r'\n+(\[.*?\]\(.*?\))', r' \1', content_md)
                content_md = re.sub(r'(\[.*?\]\(.*?\))\n+', r'\1 ', content_md)
                # 移除斜體文字前後的不必要換行
                content_md = re.sub(r'\n+(\*.*?\*)', r' \1', content_md)
                content_md = re.sub(r'(\*.*?\*)\n+', r'\1 ', content_md)
            md_out += f"### {speaker}{colon}\n{content_md}\n\n"

        elif tag == "narrative":
            kids = element_children(child)
            chosen = kids[1] if len(kids) > 1 else (kids[0] if kids else None)
            text_html = tostring(chosen) if chosen is not None else (child.text or "")
            # 對齊原始碼：narrative 的 MD 會把底線移除
            text_md = md(text_html).replace("_", "").strip()
            md_out += f"> {text_md}\n\n"

    with open(out_path, "w", encoding="utf-8") as f:
        f.write(md_out)

def main():
    if len(sys.argv) < 3:
        print(f"用法：{sys.argv[0]} input.xml output.md")
        sys.exit(1)

    in_path = sys.argv[1]
    out_path = sys.argv[2]

    # 對齊原始 Node 版本：將 & 統一替換為 &amp; 再解析
    #（若你的 XML 是有效的，可改用 recover 解析器、移除此替換）
    with open(in_path, "r", encoding="utf-8") as f:
        xml_text = f.read()
    xml_text = xml_text.replace("&", "&amp;")

    # 解析 XML（若要更寬鬆可用 XMLParser(recover=True)）
    # 將 Unicode 字串轉為 bytes 以支援 encoding declaration
    root = etree.fromstring(xml_text.encode('utf-8'))

    an2md(root, out_path)

if __name__ == "__main__":
    main()

File: an2md.ts
#!/usr/bin/env bun

import { readFileSync, writeFileSync } from "node:fs";
import { DOMParser } from "@xmldom/xmldom";

export const EN_REGEX = /^[a-zA-Z0-9$@$!%*?&#^\-_. +]+$/;
const ELEMENT_NODE = 1;
const TEXT_NODE = 3;

export function checkEn(text: string): boolean {
  return EN_REGEX.test(text ?? "");
}

function localName(node: Node | null): string {
  if (!node || node.nodeType !== ELEMENT_NODE) {
    return "";
  }
  return (node as Element).localName;
}

export function elementChildren(element: Element): Element[] {
  return [...element.childNodes].filter(
    (child): child is Element => child.nodeType === ELEMENT_NODE,
  );
}

export function sanitizeMarkdown(text: string): string {
  return text
    .replace(/\n+(\[[^\]]+\]\([^)]+\))/g, " $1")
    .replace(/(\[[^\]]+\]\([^)]+\))\n+/g, "$1 ")
    .replace(/\n+(\*[^*]+\*)/g, " $1")
    .replace(/(\*[^*]+\*)\n+/g, "$1 ")
    .trim();
}

function textContent(node: Node | null): string {
  return node?.textContent ?? "";
}

export function nodeToMarkdown(node: Node | null): string {
  if (!node) {
    return "";
  }

  if (node.nodeType === TEXT_NODE) {
    return node.nodeValue ?? "";
  }

  if (node.nodeType !== ELEMENT_NODE) {
    return "";
  }

  const element = node as Element;
  const children = [...element.childNodes].map((child) => nodeToMarkdown(child)).join("");
  switch (element.localName) {
    case "p":
      return children.trim();
    case "i":
    case "em":
      return `*${children.trim()}*`;
    case "strong":
    case "b":
      return `**${children.trim()}**`;
    case "a": {
      const href = element.getAttribute("href") ?? "";
      return `[${children.trim()}](${href})`;
    }
    case "br":
      return "\n";
    default:
      return children || textContent(element);
  }
}

function firstDescendantByLocalName(root: ParentNode, tagName: string): Element | null {
  const queue = [...root.childNodes];
  while (queue.length > 0) {
    const current = queue.shift()!;
    if (localName(current) === tagName) {
      return current as Element;
    }
    queue.push(...current.childNodes);
  }
  return null;
}

export function aknToMarkdown(xmlText: string): string {
  const sanitizedXml = xmlText.replaceAll("&", "&amp;");
  const document = new DOMParser().parseFromString(sanitizedXml, "application/xml");
  const debate = firstDescendantByLocalName(document, "debateSection");
  if (!debate) {
    throw new Error("找不到 <debateSection>");
  }

  const headingElement = firstDescendantByLocalName(debate, "heading");
  const headingText = textContent(headingElement).trim();
  const colon = checkEn(headingText) ? ":" : "：";

  let output = `# ${headingText}\n\n`;
  let lastSpeaker: string | null = null;

  for (const child of elementChildren(debate)) {
    const tagName = child.localName;
    if (tagName === "speech") {
      const speaker = (child.getAttribute("by") ?? "").replace(/^#/, "");
      const children = elementChildren(child);
      let content = "";

      if (children.length > 0) {
        const paragraphChildren = children.filter((candidate) => candidate.localName === "p");
        if (paragraphChildren.length > 0) {
          content = paragraphChildren
            .map((paragraph) => sanitizeMarkdown(nodeToMarkdown(paragraph)))
            .filter(Boolean)
            .join("\n\n");
        } else {
          const chosen = children[1] ?? children[0];
          content = sanitizeMarkdown(nodeToMarkdown(chosen));
        }
      } else {
        content = sanitizeMarkdown(textContent(child));
      }

      if (speaker === lastSpeaker) {
        output += `${content}\n\n`;
      } else {
        output += `### ${speaker}${colon}\n${content}\n\n`;
        lastSpeaker = speaker;
      }
    } else if (tagName === "narrative") {
      const children = elementChildren(child);
      const chosen = children[1] ?? children[0] ?? null;
      const narrative = sanitizeMarkdown(nodeToMarkdown(chosen)).replaceAll("_", "");
      output += `> ${narrative}\n\n`;
      lastSpeaker = null;
    }
  }

  return output;
}

export function main(argv: string[] = Bun.argv.slice(2)): number {
  if (argv.length < 2) {
    console.error(`用法：${Bun.argv[1] ?? "an2md.ts"} input.xml output.md`);
    return 1;
  }

  const [inputPath, outputPath] = argv;
  const xmlText = readFileSync(inputPath, "utf-8");
  writeFileSync(outputPath, aknToMarkdown(xmlText), "utf-8");
  return 0;
}

if (import.meta.main) {
  process.exit(main());
}
